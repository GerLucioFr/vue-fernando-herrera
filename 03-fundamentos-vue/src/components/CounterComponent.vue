<template>
  <div class="counter-container">
    <h1>{{ setTitle }}</h1>
    <p counter-id="counter">{{ counter }}{{ setPersonsCopy }}</p>
    <p>Counter value is {{ counter }}</p>

    <div>
      <button
        type="button"
        button-type="decrease"
        :disabled="counter === 0"
        @click="decreaseCounter"
      >
        Decrease -
      </button>
      <button type="button" button-type="increase" @click="increaseCounter">
        Increase +
      </button>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    title: String,
    start: {
      type: Number,
      default: 100,
      validator(value) {
        return !value || value >= 0;
      },
    },
  },
  name: 'CounterComponent',
  data() {
    return {
      counter: this.start,
    };
  },
  methods: {
    increaseCounter() {
      this.counter++;
    },
    decreaseCounter() {
      if (this.counter === 0) {
        return;
      }

      this.counter--;
    },
  },
  computed: {
    setPersonsCopy() {
      return this.counter === 0
        ? ''
        : this.counter === 1
        ? ' person'
        : ' people';
    },
    setTitle() {
      return this.title || 'Counter';
    },
  },
};
</script>

<style>
button {
  margin: 16px;
  padding: 4px 8px;

  background-color: rgb(0, 151, 58);
  color: white;
  border: none;
  border-radius: 8px;
}

button:hover {
  background-color: rgb(0, 90, 35);
}

button:disabled {
  background-color: rgb(241, 241, 241);
  color: rgb(0, 0, 0);
}
</style>
