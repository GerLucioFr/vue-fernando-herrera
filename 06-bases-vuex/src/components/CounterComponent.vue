<template>
  <div>
    <h1>Counter</h1>
    <h3>Direct access: {{ $store.state.counter.count }}</h3>
    <h3>Computed {{ countComputed }}</h3>
    <h3>Map state: {{ count }}</h3>
    <h3>Map state isLoading: {{ isLoading }}</h3>

    <h3>Direct square: {{ $store.getters['counter/getCountSquare'] }}</h3>
    <h3>Map getters: {{ getCountSquare }}</h3>

    <button @click="decrement" type="button">Decrease</button>
    <button @click="increment" type="button">Increase</button>
    <button @click="incrementBy" type="button">Increase by 5</button>
    <button :disabled="isLoading" @click="incrementRandom" type="button">
      Increase Random
    </button>
  </div>
</template>

<script>
import { mapState, mapActions, mapGetters } from 'vuex';
export default {
  name: 'CounterComponent',
  computed: {
    countComputed() {
      return this.$store.state.counter.count;
    },
    // ...mapState(['count']),
    ...mapState('counter', {
      count: 'count',
      isLoading: 'isLoading',
    }),
    ...mapGetters('counter', ['getCountSquare']),
  },
  methods: {
    increment() {
      this.$store.commit('counter/increment');
    },
    decrement() {
      this.$store.commit('counter/decrement');
    },
    incrementBy() {
      this.$store.commit('counter/incrementBy', { value: 5 });
    },
    ...mapActions('counter', ['incrementRandom']),
  },
};
</script>

<style></style>
