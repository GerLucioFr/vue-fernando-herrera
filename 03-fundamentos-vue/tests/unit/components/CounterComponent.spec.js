import CounterComponent from '@/components/CounterComponent';
import { shallowMount } from '@vue/test-utils';

describe('Counter Component Test', () => {
  let wrapper;
  let counterParagraph;
  let counterDefault;

  beforeEach(() => {
    wrapper = shallowMount(CounterComponent);
    counterParagraph = wrapper.find('[counter-id="counter"]');
  });

  it('read counter default value', () => {
    counterDefault = wrapper.props('start');

    expect(counterParagraph.text()).toBe(`${counterDefault} people`);
  });

  it('should to match with snapshot', () => {
    // expect(wrapper.html()).toMatchSnapshot();
    // Si la prueba falla por la diferencia de snapshot, pero el cambio en tu componente
    // sí es válido, vuelve a correr la prueba con la bandera -u para que actualice el
    // snapshot: npm run test:unit -u
  });

  it('title should to have default value', () => {
    const subtitle = wrapper.find('h1');
    expect(subtitle.exists()).toBe(true); // or toBeTruthy();
    expect(subtitle.text()).toBe('Counter');
  });

  it('counter should to have 100 as default value', () => {
    // const paragraphs = wrapper.findAll('p');

    expect(counterParagraph.text()).toBe('100 people');
    // paragraphs.forEach(p => {
    //   expect(Number(p.text())).toEqual(100);
    // });
  });

  it('increase button has to increment counter value by one', async () => {
    const increaseButton = wrapper.find('[button-type="increase"]');

    await increaseButton.trigger('click');
    expect(counterParagraph.text()).toBe('101 people');
  });

  it('decrease button has to decrement counter value by one', async () => {
    const decreaseButton = wrapper.find('[button-type="decrease"]');

    await decreaseButton.trigger('click');
    await decreaseButton.trigger('click');
    expect(counterParagraph.text()).toBe('98 people');
  });

  it('should to show title', () => {
    const title = 'My Title from Test';
    const tempWrapper = shallowMount(CounterComponent, {
      props: {
        title,
      },
    });

    const titleHtml = tempWrapper.find('h1');
    expect(titleHtml.text()).toBe(title);
  });
});
