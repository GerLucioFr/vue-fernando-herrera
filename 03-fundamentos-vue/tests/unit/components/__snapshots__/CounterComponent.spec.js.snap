// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component Test should to match with snapshot 1`] = `
<div class="counter-container">
  <h1>Counter</h1>
  <p counter-id="counter">100 people</p>
  <p>Counter value is 100</p>
  <div><button type="button" button-type="decrease"> Decrease - </button><button type="button" button-type="increase"> Increase + </button></div>
</div>
`;
